packages/flutter_tools/lib/src/device.dart

```diff
@@ -159,40 +166,98 @@ abstract class DeviceManager {
       Future.wait<List<Device>?>(futureDevices),
     ]);
 
+    if (waitForDeviceToConnect == true) {
+      for (final DeviceDiscovery discoverer in _platformDiscoverers) {
+        discoverer.cancelWaitForDeviceToConnect();
+      }
+    }
+
     if (exactMatchCompleter.isCompleted) {
       return <Device>[await exactMatchCompleter.future];
     }
     return prefixMatches;
   }
 
-  /// Returns the list of connected devices, filtered by any user-specified device id.
-  Future<List<Device>> getDevices() {
+  /// Returns a list of devices filtered by the user-specified device
+  /// id/name (if applicable) and [filter].
+  ///
+  /// If [filter] is not provided, a default filter that requires devices to be
+  /// connected will be used.
+  ///
+  /// (iOS devices only) If [waitForDeviceToConnect] is true and an exact match
+  /// is found but the device is not connected, wait for the device to connect.
+  Future<List<Device>> getDevices({
+    DeviceDiscoveryFilter? filter,
+    bool waitForDeviceToConnect = false,
+  }) {
     final String? id = specifiedDeviceId;
+    filter ??= DeviceDiscoveryFilter();
```

Comment:
Same nit, you can make `filter` non-nullable if you make `DeviceDiscoveryFilter` a const class:

```suggestion
    DeviceDiscoveryFilter filter = const DeviceDiscoveryFilter(),
    bool waitForDeviceToConnect = false,
  }) {
    final String? id = specifiedDeviceId;
```